src/storage/service.ts

```diff
@@ -627,33 +651,97 @@ export class StorageService {
     providers: ApprovedProviderInfo[],
     signer: ethers.Signer
   ): Promise<ApprovedProviderInfo> {
-    let randomIndex: number
+    if (providers.length === 0) {
+      throw createError(
+        'StorageService',
+        'selectRandomProvider',
+        'No providers available'
+      )
+    }
 
-    // Try crypto.getRandomValues if available (HTTPS contexts)
-    if (typeof globalThis.crypto !== 'undefined' && globalThis.crypto.getRandomValues != null) {
-      const randomBytes = new Uint8Array(1)
-      globalThis.crypto.getRandomValues(randomBytes)
-      randomIndex = randomBytes[0] % providers.length
-    } else {
-      // Fallback for HTTP contexts - use multiple entropy sources
-      const timestamp = Date.now()
-      const random = Math.random()
-      // Use wallet address as additional entropy
-      const addressBytes = await signer.getAddress()
-      const addressSum = addressBytes.split('').reduce((a, c) => a + c.charCodeAt(0), 0)
+    // Randomly sort providers inline
+    const randomlySorted: ApprovedProviderInfo[] = []
+    const remaining = [...providers]
+
+    while (remaining.length > 0) {
+      let randomIndex: number
+
+      // Try crypto.getRandomValues if available (HTTPS contexts)
+      if (typeof globalThis.crypto !== 'undefined' && globalThis.crypto.getRandomValues != null) {
+        const randomBytes = new Uint8Array(1)
+        globalThis.crypto.getRandomValues(randomBytes)
+        randomIndex = randomBytes[0] % remaining.length
+      } else {
+        // Fallback for HTTP contexts - use multiple entropy sources
+        const timestamp = Date.now()
+        const random = Math.random()
+        // Use wallet address as additional entropy
+        const addressBytes = await signer.getAddress()
+        const addressSum = addressBytes.split('').reduce((a, c) => a + c.charCodeAt(0), 0)
+
+        // Combine sources for better distribution
+        const combined = (timestamp * random * addressSum) % remaining.length
+        randomIndex = Math.floor(Math.abs(combined))
+      }
+
+      randomlySorted.push(remaining.splice(randomIndex, 1)[0])
+    }
+
+    return await StorageService.selectProviderWithPing(randomlySorted, [])
+  }
+
+  /**
+   * Select a provider from a sorted list with ping validation and exclusion support
+   * This is shared logic used by both smart selection and random selection
+   * @param sortedProviders - List of providers to try in order
+   * @param excludedProviderAddresses - List of provider addresses to exclude from selection
+   * @returns A provider that responds to ping
+   * @throws Error if no providers are reachable
+   */
+  private static async selectProviderWithPing (
+    sortedProviders: ApprovedProviderInfo[],
+    excludedProviderAddresses: string[]
+  ): Promise<ApprovedProviderInfo> {
+    // Filter out excluded providers
+    const availableProviders = sortedProviders.filter(p => !excludedProviderAddresses.includes(p.owner))
```

Comment:
```suggestion
  /**
   * Select a provider from a sorted list with ping validation.
   * This is shared logic used by both smart selection and random selection.
   * @param sortedProviders - List of providers to try in order
   * @returns A provider that responds to ping
   * @throws Error if no providers are reachable
   */
  private static async selectProviderWithPing (sortedProviders: ApprovedProviderInfo[]): Promise<ApprovedProviderInfo> {
```

We're just not using the exclusion feature here are we? Let's ditch it, and from the tests too. Unless I'm missing something?